internal/controller/logger_manager.go

```diff
@@ -0,0 +1,267 @@
+// Copyright 2024 Illumio, Inc. All Rights Reserved.
+
+package controller
+
+import (
+	"io"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	pb "github.com/illumio/cloud-operator/api/illumio/cloud/k8scluster/v1"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+const (
+	maxBufferSize = 100
+	flushInterval = 5 * time.Second
+)
+
+// BufferedGrpcWriteSyncer is a custom zap writesync that writes to a grpc stream
+// In case stream is not connected it will buffer to memory
+type BufferedGrpcWriteSyncer struct {
+	client    pb.KubernetesInfoService_SendLogsClient
+	conn      *grpc.ClientConn
+	buffer    []zapcore.Entry
+	mutex     sync.Mutex
+	done      chan struct{}
+	logger    *zap.SugaredLogger
+	logLevel  *zap.AtomicLevel
+	encoder   zapcore.Encoder
+	lostLogs  bool
+	lostBytes int
+}
+
+// NewBufferedGrpcWriteSyncer returns a new BufferedGrpcWriteSyncer
+func NewBufferedGrpcWriteSyncer(client pb.KubernetesInfoService_SendLogsClient, conn *grpc.ClientConn) *BufferedGrpcWriteSyncer {
+	bws := &BufferedGrpcWriteSyncer{
+		client:   client,
+		conn:     conn,
+		buffer:   make([]zapcore.Entry, 0, maxBufferSize),
+		done:     make(chan struct{}),
+		lostLogs: false,
+	}
+	go bws.run()
+	return bws
+}
+
+// Sync flushes buffered log data into grpc stream if possible.
+func (b *BufferedGrpcWriteSyncer) Sync() error {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+
+	b.flush()
+	close(b.done)
+	return b.conn.Close()
+}
+
+// flush will attempt to dump buffer into GRPC stream if available
+func (b *BufferedGrpcWriteSyncer) flush() {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+
+	if len(b.buffer) == 0 || b.conn == nil || b.conn.GetState() != connectivity.Ready {
+		if len(b.buffer) > maxBufferSize {
+			b.lostLogs = true
+			b.lostBytes += maxBufferSize - len(b.buffer)
+		}
+		return
+	}
+
+	for _, logEntry := range b.buffer {
+		if err := b.sendLog(logEntry); err != nil {
+			b.logger.Errorw("Failed to send log",
+				"error", err,
+				zap.Any("logBufferSize", len(b.buffer)),
+			)
+			b.lostLogs = true
+			b.lostBytes += len(logEntry.Message)
+			return
```

Comment:
You can't just return here, since you may have successfully sent some log entries before starting to fail.
And those entries are still in the buffer. You must remove them from the buffer before returning.